Add release-apt-get workflow

ldennington opened the pull request:
Adding workflow to microsoft/git to release to Microsoft's official repositories for Ubuntu distributions we currently support: `HIRSUTE` (21.04, aka latest) and `BIONIC` (18.04, aka latest LTS). Details for the reasons behind this change can be found [here](https://github.com/github/git-client/issues/370#issuecomment-857209503).

 **Note:** This is the same flow you can already find in [GCM Core](https://github.com/microsoft/Git-Credential-Manager-Core/blob/main/.github/workflows/release-apt-get.yml), minus signing, since our build takes care of that on our hehalf. I've also [validated this flow](https://github.com/ldennington/git/actions/runs/951387139) in my fork to create successful [bionic](https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod/pool/main/g/git-vfs/) and [hirsute](https://packages.microsoft.com/repos/microsoft-ubuntu-hirsute-prod/pool/main/g/git-vfs/) releases.

derrickstolee commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,52 @@
+  workflow_dispatch:
+    inputs:
+      release:
+        description: 'Release Tag'
+        required: true
+        default: 'latest'
```
Just so I understand, this workflow triggers when we publish a release that becomes the "latest" release. We don't really use any information about that release from the workflow startup details because `fetch_release.py` will download from the public feed and push whatever is found there.

I'm not sure if this indirection is _necessary_ or if there is a "blessed" way to do it in GitHub Actions. I don't care that much as long as it works (and I understand your approach here).

ldennington replied:
This is a special (and very handy!) action that allows us to run the workflow manually. I can't give an example from microsoft/git yet, as the code has not made it into the default branch. However, here's what this looks like in GCM Core:

![image](https://user-images.githubusercontent.com/11321782/122833339-a6816e80-d2a1-11eb-86c5-dd34aa784361.png)

This will not impact the workflows that trigger automatically when a new release becomes available.

ldennington replied:
Here is some further reading if you'd like to learn more!
https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow

derrickstolee reviewed (approved):
I read through the code and nothing looks strange, other than the things that seem like they are _required_ to look strange. There's an interesting indirection that I mention in a comment, but don't spend too much time thinking about that.

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,52 @@
+        RELEASE: ${{ github.event.inputs.release }}
+      run: |
+        pip install requests
+        python .github/fetch_release.py
```
It might make sense to turn this into a [`github-script`](https://github.com/marketplace/actions/github-script) task instead: that allows you to use the REST API without much hassle and with less code. You can probably copy-edit what I did [in  `setup-git-for-windows-sdk`](https://github.com/git-for-windows/setup-git-for-windows-sdk/blob/e377dda38b797976b20f6dd45098f306d5b0b137/.github/workflows/release-tag.yml#L64-L81).

ldennington replied:
Sure, I'm fine with that. I'm just having a little trouble figuring out what [this piece](https://github.com/git-for-windows/setup-git-for-windows-sdk/blob/e377dda38b797976b20f6dd45098f306d5b0b137/.github/workflows/release-tag.yml#L73-L81) is using. Is there a corresponding method for getting a release (I assume so, but want to be sure)? Would you mind pointing me to any existing docs? I've done a couple searches but am having some trouble getting a matching result.

dscho replied:
I pushed an example demonstrating this idea to my fork: https://github.com/dscho/git/actions/runs/966065373/workflow

dscho replied:
The essence of this example is this step:

```yml
      - name: "Get release assets"
        id: get-asset
        env:
          RELEASE: ${{ github.event.inputs.release }}
        uses: actions/github-script@v3
        with:
          github-token: ${{secrets.GITHUB_TOKEN}}
          script: |
            const { data } = await github.repos.getRelease({
                owner: context.repo.owner,
                repo: context.repo.repo,
                release_id: process.env.RELEASE || 'latest'
              })
            const assets = data.assets.filter(asset => asset.name.endsWith('.deb'))
            if (assets.length !== 1) {
              throw new Error(`Unexpected number of .deb assets: ${assets.length}`)
            }
            core.setOutput('name', assets[0].name)
            core.setOutput('url', assets[0].url)
```

In subsequent steps, you can use `${{ steps.get-asset.outputs.name }}` and `${{ steps.get-asset.outputs.url }}` to refer to the discovered asset name/URL.

BTW I could imagine that things would look a bit simpler if you used another `github-script` step to generate and write the JSON configuration. Something along these lines:

```yml
      - name: "Write repoclient config"
        env:
          AZURE_AAD_ID: ${{ secrets.AZURE_AAD_ID }}
          AAD_CLIENT_SECRET: ${{ secrets.AAD_CLIENT_SECRET }}
        uses: actions/github-script@v3
        with:
          github-token: ${{secrets.GITHUB_TOKEN}}
          script: |
            for (const key of ['AZURE_AAD_ID', 'AAD_CLIENT_SECRET']) {
              if (!process.env[key]) throw new Error(`Required env var ${key} is missing!`)
            }

            const config = {
              AADResource: 'https://microsoft.onmicrosoft.com/945999e9-da09-4b5b-878f-b66c414602c0',
              AADTenant: '72f988bf-86f1-41af-91ab-2d7cd011db47',
              AADAuthorityUrl: 'https://login.microsoftonline.com',
              server: 'azure-apt-cat.cloudapp.net',
              port: '443',

              AADClientId: process.env.AZURE_AAD_ID,
              AADClientSecret: process.env.AAD_CLIENT_SECRET,
              repositoryId: ''
            }

            const fs = require('fs')
            fs.writeFileSync('config.json', JSON.stringify(config, null, 2))
```

Technically, this is "abusing" the `github-script` Action a bit because we are not even attempting to access any GitHub resources. But practically it is a really convenient way to execute Javascript snippets.

dscho replied:
By the way, I should clarify how I developed this example workflow: I _definitely_ did not edit-and-push the workflow all the time until it worked (that would take _way_ too much time, unfortunately debugging facilities are still a weak point of pretty much all CI systems out there).

Instead, I started in an empty directory, called `npm init` in there and then `npm install @octokit/rest`. Then, I wrapped the snippet in some boiler plate and saved it as a `.js` file:

```js
const octokit = require('@octokit/rest');
const github = new octokit.Octokit();

const context = {
  repo: {
    owner: 'microsoft',
    repo: 'git'
  }
};

const core = {
  setOutput: (key, value) => {
    console.log(`Setting output ${key} to ${value}`)
  }
};

(async () => {
  [... snippet here ...]
})().catch(e => console.log(e));
```

And then I ran this via `node example.js` in a terminal while having the file open in a VS Code instance, for some rapid edit/run cycle.

Another favorite GitHub Actions debugging technique of mine involves running [`act`](https://github.com/nektos/act) locally. The biggest challenge with that is that it requires Docker, and the Linux variant at that (which also means that you can only simulate workflow jobs that target a Linux agent, but that is frequently good enough for me). To accommodate that, I installed Docker in my WSL instance, which works well enough (but it did have to download almost half a Maven repository worth of bytes first, if you know what I mean).

ldennington replied:
Thank you so much for the detailed description, @dscho! This will be invaluable to my GitHub Actions flows far into the future :).

dscho commented on .github/configure_repoclient.py:
```diff
@@ -0,0 +1,33 @@
+	("config.json", repo_config),
+]
+
+for filename, data in configs:
```
Unless we're planning on writing more files than just `config.json`, maybe we should skip that `for` loop and the `configs` array?

ldennington replied:
Sure, that's fine.

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,52 @@
+      env:
+        AZ_SUB: ${{ secrets.AZURE_SUBSCRIPTION }}
+      run: |
+        az storage blob download --subscription  "$AZ_SUB" --account-name gitcitoolstore -c tools -n azure-repoapi-client_2.0.1_amd64.deb -f repoclient.deb --auth-mode login
```
Is there no "more official" place to get this Debian package? I worry about hard-coding the version number "2.0.1" here...

ldennington replied:
There is a way to get it from http://tux-devrepo.corp.microsoft.com, but I'm going to follow up with Kyle. I'm sure he likely had a good reason for not going that route when he set this up initially.

dscho replied:
I just don't want to be stuck with a stale version and only realizing when something hits the fan that we simply forgot to upgrade ;-)

ldennington replied:
Ok, so the issue is that the tool is stored on http://tux-devrepo.corp.microsoft.com, which is inaccessible to machines that are not connected to corpnet. Kyle recommends setting up an internal pipeline to download latest and publish to blob storage. I have opened [this issue](https://github.com/github/git-client/issues/467) accordingly.

dscho reviewed (approved):
👨‍🍳 💋

dscho committed:
Add workflow for apt-get release

This adds support for releasing to Ubuntu repositories hosted
at http://packages.microsoft.com/ (hosting location for Microsoft's
official apt/yum repos). This allows users to install via apt-get on
Hirsute/Bionic. Details to configure appropriate repos can be found
here:

https://docs.microsoft.com/en-us/windows-server/administration/Linux-Package-Repository-for-Microsoft-Software).

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,106 @@
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - name: setup python
```
Once we write the JSON via Javascript, we won't need this, right?

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,106 @@
+      - name: "Install Repo Client"
+        env:
+          AZURE_AAD_ID: ${{ secrets.AZURE_AAD_ID }}
+          AAD_CLIENT_SECRET: ${{ secrets.AAD_CLIENT_SECRET }}
```
My mistake: this `env` block is needed in the step below, "Configure Repo Client".

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,106 @@
+        run: |
+          sudo apt-get install python3-adal --yes
+          sudo dpkg -i repoclient.deb
+          python .github/configure_repoclient.py
```
This is now done below, in the `github-script` step, right?

ldennington replied:
I was able to remove the reference to the old script (thanks for calling that out!). However, we still need to install `pythin3-adal`, as it's a repoclient dependency.

dscho replied:
Okay, thanks for explaining!

dscho commented on .github/workflows/release-apt-get.yml:
```diff
@@ -0,0 +1,106 @@
+        env:
+          RELEASE: ${{ github.event.inputs.release }}
+        run: |
+          pip install requests
```
Do we still need this?

ldennington replied:
Nope!

dscho reviewed (commented):
Just a few things I noticed.
